### PR TITLE
#166789989 Users should receive a JWT on Successful Registration.

### PR DESCRIPTION
**Pull request by elemanhillary-zz**
#### What does this PR do?
- Adds jsonwebtoken strategy for authentication
- Adds registration and login endpoint

#### Description of Task to be completed
- Enable user to register and login
- Enable user authentication with jwt(jsonwebtoken) authentication
#### How should this be manually tested?
1. Run the following commands in Terminal
```
git clone https://github.com/andela/ah-92explorers-backend.git
cd ah-92explorers-backend
sudo mv .env.example .env
npm install
sequelize db:migrate
npm start
```
2. Open your favourite browser and navigate to the following:
```
localhost:3000/api/users
localhost:3000/api/users/login
```

#### Any background context you want to provide?
- None

#### What are the relevant pivotal tracker stories?
[166789989](https://www.pivotaltracker.com/story/show/166789989)

#### Screenshots (if appropriate)
Imgur

#### Checklist

- [x] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I don't have more than 3 major commits on this PR

### CODE REVIEW

**Review by kodek-sleuth — CHANGES_REQUESTED** [review]
Please don't return the success key when returning an error even if its 0
Please use message key not msg

File: tests/auth.test.js [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+import chaiHttp from 'chai-http';
+import faker from 'faker';
+import app from '../index';
+
+chai.use(chaiHttp);
+
+const email = faker.internet.email();
+const userName = faker.internet.userName();
+
+const dummyData = {
```
Comment (kodek-sleuth) [inline tests/auth.test.js]: we are going to use seeders for storing dummy data

**Review by kodek-sleuth — CHANGES_REQUESTED** [review]
Please don't return a password/id in the response and I would prefer if we used UUID to generate id's

**Review by niyoceles — DISMISSED** [review]
The test is not passing on Travis CI, please rework on it.

**Comment by elemanhillary-zz** [comment]
> The test is not passing on Travis CI, please rework on it.

Tests are now passing @niyoceles

File: .travis.yml [hunk 2]
```diff
@@ -1,7 +1,11 @@
...
+node_js:
   - "stable"
 services:
   - postgresql
+before_script:
+  - psql -c 'CREATE DATABASE users_test;' -U postgres
+  - npm i sequelize -g
+  - sequelize db:migrate
 after_success:
-  - npm test
```
Comment (kodek-sleuth) [inline .travis.yml]: This command should be run in the script not here. It completely overlooks the presence of the latter

**Review by kodek-sleuth — CHANGES_REQUESTED** [review]

File: migrations/20190704183748-create-user.js [hunk 3]
```diff
@@ -0,0 +1,39 @@
+module.exports = {
```
Comment (kodek-sleuth) [inline migrations/20190704183748-create-user.js]: Use es6 here, @slack here

**Review by kodek-sleuth — CHANGES_REQUESTED** [review]

**Review by kodek-sleuth — CHANGES_REQUESTED** [review]
Please follow all eslint rules

File: .travis.yml (hunk 2 above)
Comment (elemanhillary-zz) [reply .travis.yml]: @kodek-sleuth kindly learn how to review, code which is in a red line indicates it has been deleted and changed with a new piece of code.

**Review by IsaiahRn — DISMISSED** [review]
You have implemented authentication and jwt strategy for authentication

File: .editorconfig [hunk 4]
```diff
@@ -0,0 +1,8 @@
+root = true
```
Comment (elemanhillary-zz) [inline .editorconfig]: remove `.editorconfig` from version control

File: .eslintignore [hunk 5]
```diff
@@ -0,0 +1,2 @@
+node_modules
+!.eslintrc
```
Comment (elemanhillary-zz) [inline .eslintignore]: remove `!.eslintrc`

File: .eslintrc [hunk 6]
```diff
@@ -15,14 +15,14 @@
         "sourceType": "module"
```
Comment (elemanhillary-zz) [inline .eslintrc]: use `0` to turn off some eslint rules

File: controllers/auth.js [hunk 7]
```diff
@@ -0,0 +1,60 @@
+import db from '../models';
+import Auth from '../helpers/auth';
+
+export const signup = async (req, res) => {
+  const {
+    username, firstName, lastName, email,
+  } = req.body;
+  const password = Auth.hashPassword(req.body.password);
```
Comment (elemanhillary-zz) [inline controllers/auth.js]: 1. destructure password in the request body on line `6`
2. rename `password` to `hashedPassword`

File: controllers/auth.js [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+  const transaction = await db.sequelize.transaction();
+  try {
+    const newUser = await db.User.create({
+      username, firstName, lastName, email, password,
+    }, {
+      transaction,
+    });
+    await transaction.commit();
+    if (newUser) {
+      return res.status(201).json({
```
Comment (elemanhillary) [inline controllers/auth.js]: restructure:
```
{
message,
   user: {
       token,
       username,
       email,
     }
}
```

File: controllers/auth.js [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+        success: 1,
+        token: Auth.genToken(username, firstName, lastName, email),
```
Comment (elemanhillary-zz) [inline controllers/auth.js]: change token to take these params
```
Auth.genToken(username, email)
```

File: controllers/auth.js [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+};
+
+export const signin = async (req, res) => {
+  const { email, password } = req.body;
+  const transaction = await db.sequelize.transaction();
+  try {
+    const user = await db.User.findOne({ where: { email } }, { transaction });
+    await transaction.commit();
+    if (!user) {
+      return res.status(401).json({ message: 'user doesnot exit' });
```
Comment (elemanhillary-zz) [inline controllers/auth.js]: fix typos `exit` to `exist`

File: controllers/auth.js [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+    }
+    const passBool = Auth.comparePassword(password, user.password);
+    if (passBool) {
+      return res.status(200).json({
+        success: 1,
+        token: Auth.genToken(user.username, user.firstName, user.lastName, user.email),
```
Comment (elemanhillary-zz) [inline controllers/auth.js]: destructure user object return by the db

File: helpers/auth.js [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+class Auth {
+  static hashPassword(password) {
+    return bcrypt.hashSync(password, bcrypt.genSaltSync(10));
+  }
+
+  static comparePassword(password, hashedPassword) {
+    return bcrypt.compareSync(password, hashedPassword);
+  }
+
+  static genToken(username, firstname, lastname, email) {
```
Comment (elemanhillary-zz) [inline helpers/auth.js]: genToken shoud take `username, email` as params

File: middlewares/checkToken.js [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+          message: error.message,
+        });
+      }
+      req.decoded = decoded;
+      next();
+    });
+  } else {
+    return res.status(401).json({
+      success: false,
+      message: 'token empty',
```
Comment (elemanhillary-zz) [inline middlewares/checkToken.js]: change `token empty` to `missing authorization token` and json key` message` to `error`

File: middlewares/checkToken.js [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+          success: false,
+          message: error.message,
+        });
+      }
+      req.decoded = decoded;
+      next();
+    });
+  } else {
+    return res.status(401).json({
+      success: false,
```
Comment (kodek-sleuth) [inline middlewares/checkToken.js]: If you sending an error key, we should not use a success key

**Review by kodek-sleuth — CHANGES_REQUESTED** [review]

File: routes/api/index.js [hunk 15]
```diff
@@ -1,8 +1,8 @@
 import express from 'express';
-import welcome from './welcome';
+import routes from './users';
 
 const router = express.Router();
 
-router.use('/', welcome);
+router.use('/', routes);
```
Comment (kodek-sleuth) [inline routes/api/index.js]: please use users instead of routes

**Review by kodek-sleuth — DISMISSED** [review]

**Review by IsaiahRn — DISMISSED** [review]
It can be merged!

File: .gitignore [hunk 16]
```diff
@@ -1,10 +1,11 @@
 # Logs
 logs
 *.log
+.dist
 .DS_Store
 
 npm-debug.log*
-
+seeders
```
Comment (kodek-sleuth) [inline .gitignore]: Remove seeders from the gitignore file

**Review by kodek-sleuth — APPROVED** [review]

File: .gitignore (hunk 16 above)
Comment (kodek-sleuth) [inline .gitignore]: Remove seedeers from the gitignore file

**Review by kodek-sleuth — CHANGES_REQUESTED** [review]

File: .sequelizerc [hunk 17]
```diff
@@ -1,5 +1,4 @@
 require('@babel/register');
-
 const path = require('path');
 
 module.exports = {
```
Comment (kodek-sleuth) [inline .sequelizerc]: Use es6 to export this

**Review by kodek-sleuth — CHANGES_REQUESTED** [review]

File: controllers/auth.js [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+    await transaction.rollback();
+    return res.status(409).json({ message: `${ex.errors[0].path.toLowerCase()} already exists` });
+  }
+};
+
+export const signin = async (req, res) => {
+  const { email, password } = req.body;
+  const transaction = await db.sequelize.transaction();
+  try {
+    const user = await db.users.findOne({ where: { email: req.body.email } }, { transaction });
```
Comment (kodek-sleuth) [inline controllers/auth.js]: This users object can be destructured from the db object you imported e.g 

`import db from ../models`;
const { users } = db;

**Review by kodek-sleuth — CHANGES_REQUESTED** [review]

File: tests/token.test.js [hunk 19]
```diff
@@ -0,0 +1,51 @@
...
+    next = sinon.spy();
+  });
+  it('next should not be called if bad token was provided', () => {
+    request.headers = {};
+    request.headers.authorization = 'some authorization header';
+    checkToken(request, response, next);
+    expect(next.called).to.equal(false);
+  });
+
+  it('request should contain user info if good token was provided', () => {
```
Comment (kodek-sleuth) [inline tests/token.test.js]: begin your it blocks with should

**Review by kodek-sleuth — DISMISSED** [review]

**Review by niyoceles — DISMISSED** [review]

**Review by IsaiahRn — DISMISSED** [review]
It is ready to be merged!

File: controllers/auth.js [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+    await transaction.commit();
+    if (!user) {
+      return res.status(404).json({ message: 'user doesnot exist' });
```
Comment (kodek-sleuth) [inline controllers/auth.js]: I think you should re-edit this message to `user not found`

**Review by kodek-sleuth — DISMISSED** [review]

File: controllers/auth.js [hunk 21]
```diff
@@ -0,0 +1,61 @@
...
+import Auth from '../helpers/auth';
+
+const { users } = db;
+
+export const signup = async (req, res) => {
+  const {
+    username, firstname, lastname, email, password,
+  } = req.body;
+  const hashedPassword = Auth.hashPassword(password);
+  const transaction = await db.sequelize.transaction();
```
Comment (kodek-sleuth) [inline controllers/auth.js]: restructure this line please

**Review by kodek-sleuth — CHANGES_REQUESTED** [review]

**Review by kodek-sleuth — APPROVED** [review]
This branch is good to be merged

**Review by niyoceles — APPROVED** [review]

**Review by IsaiahRn — APPROVED** [review]
It can be merged!

**Pull request merged**